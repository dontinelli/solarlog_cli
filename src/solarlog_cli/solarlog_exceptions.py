"""Exceptions for Python client for Solar-Log."""


class SolarLogError(Exception):
    """Generic exception."""

class SolarLogConnectionError(SolarLogError):
    """Analytics connection exception."""

class SolarLogUpdateError(SolarLogError):
    """Analytics connection exception."""
