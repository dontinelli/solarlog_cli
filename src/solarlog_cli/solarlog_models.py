"""Models for SolarLog."""
from dataclasses import dataclass, field
from datetime import datetime

from mashumaro import DataClassDictMixin

@dataclass
class InverterData():
    """Inverter Data model."""

    name: str = ""
    enabled: bool = False
    current_power: float | None = None
    consumption_year: float | None= None


@dataclass
class SolarlogData(DataClassDictMixin):
    """Basic Data model."""

    # pylint: disable=too-many-instance-attributes

    consumption_ac: float
    consumption_day: float
    consumption_month: float
    consumption_total: float
    consumption_yesterday: float
    consumption_year: float
    last_updated: datetime
    power_ac: float
    power_dc: float
    total_power: float
    voltage_ac: float
    voltage_dc: float
    yield_day: float
    yield_yesterday: float
    yield_month: float
    yield_year: float
    yield_total: float

    #calculated values
    alternator_loss: float = 0
    capacity: float | None = None
    efficiency: float | None = None
    power_available: float = 0
    usage: float | None = None

    #extended data
    inverter_data: dict[int, InverterData] = field(default_factory=dict)
    production_year: float | None = None
    self_consumption_year: float | None = None
