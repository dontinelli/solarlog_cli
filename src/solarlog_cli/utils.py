"""Utility functions for the FYTA integration."""


def safe_get(plant_data, key_path, expected_type):
    """Get a value from a nested dictionary and cast it to the expected type.

    Args:
        plant_data: The dictionary to get the value from.
        key_path: The path to the value in the dictionary. Se√ºarated by dots.
        expected_type: The expected type of the value (int, float, bool and str supported).

    Returns: The value from the dictionary or None if the value is not found or the type is not
    as expected.
    """
    keys = key_path.split(".")
    value = plant_data
    try:
        for key in keys:
            value = value.get(key)
            if value is None:
                return None
    except AttributeError:
        return None
    return_value = None
    if expected_type == int:
        return_value = __get_int(value)
    if expected_type == float:
        return_value =  __get_float(value)
    if expected_type == bool:
        return_value = __get_bool(value)
    if expected_type == str:
        return_value = str(value)
    return return_value


def __get_bool(value):
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        if value.lower() == "true":
            return True
        if value.lower() == "false":
            return False
    return None


def __get_float(value):
    try:
        return float(value)
    except (ValueError, TypeError):
        return None


def __get_int(value):
    try:
        return int(value)
    except (ValueError, TypeError):
        return None
