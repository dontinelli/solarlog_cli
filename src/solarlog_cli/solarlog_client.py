"""Client to access Solar-Log."""

from __future__ import annotations

import asyncio
from datetime import datetime
import json
import logging
from typing import Any

from aiohttp import ClientSession

from .solarlog_exceptions import SolarLogConnectionError, SolarLogUpdateError

SOLARLOG_REQUEST_PAYLOAD = {801: {170: None}}

_LOGGER = logging.getLogger(__name__)


class Client:
    """Client class to access Solar-Log."""

    def __init__(self, host: str):
        self.host: str = host

        self.session: ClientSession = ClientSession()
        self._close_session: bool = True

        self.request_timeout = 10

    async def test_connection(self) -> bool:
        """Test the connection to Solar-Log"""

        url = f"{self.host}/getjp"

        response = await self.session.post(url, json=SOLARLOG_REQUEST_PAYLOAD)

        if response.status == 200:
            return True

        return False

    async def execute_http_request(self, json_payload) -> dict[str, Any]:
        """Hepler function to process the HTTP Get call."""
        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        url = f"{self.host}/getjp"

        header = {
            "Content-Type": "application/json",
        }

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.post(
                    url=url, headers=header, json=json_payload
                )
        except asyncio.TimeoutError as exception:
            msg = f"Timeout occurred while connecting to Solar-Log at {self.host}"
            raise SolarLogConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if response.status != 200:
            msg = f"The server responded with error code {response.status} while fetching data from Solar-Log at {self.host}.\n{url}\n{header}\n{json_payload}"
            text = await response.text()
            raise SolarLogUpdateError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        text = await response.text()

        try:
            json_response = json.loads(text)
        except ValueError as err:
            msg = f"Value error while decoding response: {err}."
            raise SolarLogUpdateError(
                msg,
                {"Content-Type": content_type, "Server response": text},
            ) from err

        return json_response

    async def get_basic_data(self) -> dict[str, Any]:
        """Get basic data from Solar-Log."""

        raw_data: dict = await self.execute_http_request({801: {170: None}})
        raw_data = raw_data["801"]["170"]

        data = {
            "last_updated": datetime.strptime(raw_data["100"], "%d.%m.%y %H:%M:%S"),
            "power_ac": raw_data["101"],
            "power_dc": raw_data["102"],
            "voltage_ac": raw_data["103"],
            "voltage_dc": raw_data["104"],
            "yield_day": raw_data["105"],
            "yield_yesterday": raw_data["106"],
            "yield_month": raw_data["107"],
            "yield_year": raw_data["108"],
            "yield_total": raw_data["109"],
            "consumption_ac": raw_data["110"],
            "consumption_day": raw_data["111"],
            "consumption_yesterday": raw_data["112"],
            "consumption_month": raw_data["113"],
            "consumption_year": raw_data["114"],
            "consumption_total": raw_data["115"],
            "total_power": raw_data["116"],
        }

        return data

    async def get_power_per_inverter(self) -> dict[int, float]:
        """Get power data from Solar-Log"""

        raw_data: dict = await self.execute_http_request({782: None})

        data = {int(key): val for key, val in raw_data["782"].items() if val != "0"}

        return data

    async def get_energy_per_inverter(self) -> dict[int, float]:
        """Get power data from Solar-Log"""

        raw_data: dict = await self.execute_http_request({854: None})
        data_list = raw_data["854"][-1][-1]

        data = {}

        for item in data_list:
            if item != 0:
                data |= {int(data_list.index(item)): item}

        return data

    async def get_energy(self) -> dict[str, float]:
        """Get power data from Solar-Log"""

        raw_data: dict = await self.execute_http_request({878: None})

        data = {
            "production_year": raw_data["878"][-1][1],
            # "consumption_year": raw_data["878"][-1][2], # already in basic data
            "self_consumption_year": raw_data["878"][-1][3],
        }

        return data

    async def get_device_list(self) -> dict[int, dict[str, str]]:
        """Get list of all connected devices."""

        # get list of all inverters connected to Solar-Log
        raw_data: dict = await self.execute_http_request({740: None})
        raw_data = raw_data["740"]

        device_list: dict[int, dict[str, str]] = {}

        for key, value in raw_data.items():
            if value != "Err":
                # get name of the inverter
                raw_data = await self.execute_http_request({141: {key: {119: None}}})
                device_list |= {int(key): {"name": raw_data["141"][key]["119"]}}

        return device_list

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()
