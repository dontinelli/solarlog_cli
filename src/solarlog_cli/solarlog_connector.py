"""Connector class to manage access to Solar-Log."""

from datetime import timezone
from typing import Any
from zoneinfo import ZoneInfo

from .solarlog_client import Client
from .solarlog_exceptions import SolarLogUpdateError


class SolarLogConnector:
    """Connector class to access Solar-Log."""

    def __init__(
        self,
        host: str,
        extended_data: bool = False,
        tz: str = "",
        device_enabled: dict[int, bool] | None = None,
    ):
        self.client = Client(host)
        self.extended_data: bool = extended_data
        self._device_enabled: dict[int, bool] | None = device_enabled
        self._device_list: dict[int, dict[str, str | bool]] = {}

        self.timezone: ZoneInfo = timezone.utc if tz == "" else ZoneInfo(tz)

    async def test_connection(self) -> bool:
        """Test if connection to Solar-Log works."""

        return await self.client.test_connection()

    async def update_data(self) -> dict[str, Any]:
        """Get data from Solar-Log."""

        data: dict = await self.client.get_basic_data()

        if data["last_updated"].year == 1999:
            raise SolarLogUpdateError(
                "Invalid data returned (can happen after Solarlog restart)."
            )

        data["last_updated"] = data["last_updated"].astimezone(self.timezone)

        # print(f"basic data updated: {data}")
        if self.extended_data:
            data |= await self.client.get_energy()
            data |= {"devices": await self.update_inverter_data()}
            # print(f"extended data updated: {data}")
        return data

    async def update_device_list(self) -> dict[int, dict[str, str | bool]]:
        """Update list of devices."""
        if not self.extended_data:
            return {}

        device_list = await self.client.get_device_list()

        for key, value in self._device_enabled.items():
            device_list[int(key)] |= {"enabled": value}
        # print(f"device list: {device_list}")
        self._device_list = device_list

        return device_list

    async def update_inverter_data(self) -> dict[int, dict[str, Any]]:
        """Update device specific data."""
        data: dict[int, dict[str, Any]] = {}
        # print(self._device_enabled)
        raw_data = await self.client.get_power_per_inverter()
        # print(f"power: {raw_data}")
        for key, value in raw_data.items():
            key = int(key)
            if self._device_enabled[key]:
                data |= {key: {"current_power": float(value)}}

        raw_data = await self.client.get_energy_per_inverter()
        # print(f"energy: {raw_data}")
        for key, value in raw_data.items():
            # print(f"key: {key}, value: {value}, enabled: {self._device_enabled[key]}")
            if self._device_enabled[key]:
                if key in data:
                    data[key].update({"consumption_year": float(value)})
                else:
                    data |= {key: {"consumption_year": float(value)}}
        # print(data)

        return data

    @property
    def host(self) -> str:
        """Host of Solar-Log."""
        return self.client.host

    @property
    def device_list(self) -> dict[int, dict[str, str | bool]]:
        """List of all devices of Solar-Log."""
        return self._device_list

    def device_name(self, device_id: int) -> str:
        """Get name of Solar-Log attached device."""
        print(f"list: {self._device_list}, id: {device_id}")
        if device_id in self._device_list:
            return self._device_list[device_id]["name"]

        return ""

    def device_enabled(self, device_id: int = None) -> bool | dict[int, bool]:
        """Status of inverter."""
        if device_id is None:
            return self._device_enabled
        return self._device_enabled[device_id]
