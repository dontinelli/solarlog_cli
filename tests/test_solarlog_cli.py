"""Tests for solarlog_cli."""

from aioresponses import aioresponses

import pytest
from syrupy.assertion import SnapshotAssertion

from solarlog_cli.solarlog_connector import SolarLogConnector
from solarlog_cli.solarlog_exceptions import (
    SolarLogAuthenticationError,
    SolarLogConnectionError,
    SolarLogError,
    SolarLogUpdateError,
)

from . import load_fixture


@pytest.mark.parametrize(
    ("response_status", "return_value"),
    [
        (200, True),
        (400, False),
    ],
)
async def test_connection(
    responses: aioresponses,
    response_status: str,
    return_value: bool,
) -> None:
    """Test connection."""
    responses.post(
        "localhost/getjp",
        status=response_status,
    )

    solarlog_connector = SolarLogConnector("localhost")

    await solarlog_connector.client.close()
    solarlog_connector.client.session = None  # type: ignore [assignment]
    assert await solarlog_connector.test_connection() == return_value

    assert solarlog_connector.host == "localhost"

    assert solarlog_connector.client.session is not None
    assert not solarlog_connector.client.session.closed
    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed


async def test_extended_data_available(
    responses: aioresponses,
) -> None:
    """Test extended data available."""

    solarlog_connector = SolarLogConnector("http://solarlog.com", password="pwd")

    responses.post(
        "http://solarlog.com/getjp",
        timeout=True,
    )
    assert not await solarlog_connector.test_extended_data_available()

    responses.post(
        "http://solarlog.com/getjp",
        status=400,
    )
    assert not await solarlog_connector.test_extended_data_available()

    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_list_access_denied.json"),
    )
    responses.post(
        "http://solarlog.com/login",
        body="FAILED - User was wrong",
    )
    assert not await solarlog_connector.test_extended_data_available()
    assert solarlog_connector.password == ""

    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_list.json"),
    )
    assert await solarlog_connector.test_extended_data_available()

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed

async def test_login_and_data_retreival(responses: aioresponses) -> None:
    """Test login into Solar-Log."""
    responses.post(
        "http://solarlog.com/login",
        headers={"Set-Cookie": "SolarLog=token"},
        body="SUCCESS - Password was correct, you are now logged in",
    )
    solarlog_connector = SolarLogConnector("http://solarlog.com", password="pwd")

    assert await solarlog_connector.login()
    assert solarlog_connector.client.token == "token"

    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("basic_data.json"),
    )
    await solarlog_connector.update_data()

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed

async def test_login_exceptions(responses: aioresponses) -> None:
    """Test exceptions at login into Solar-Log."""
    solarlog_connector = SolarLogConnector("http://solarlog.com", password="pwd")

    responses.post(
        "http://solarlog.com/login",
        body="FAILED - Password was wrong",
    )
    with pytest.raises(SolarLogAuthenticationError):  # type: ignore [call-overload]
        await solarlog_connector.client.login()

    responses.post(
        "http://solarlog.com/login",
        body="FAILED - User was wrong",
    )
    assert not await solarlog_connector.client.login()

    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_list.json"),
    )
    assert not await solarlog_connector.client.login()

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed

async def test_update_data(
    responses: aioresponses,
    snapshot: SnapshotAssertion
) -> None:
    """Test update data."""
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("basic_data.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("extended_data.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("power_per_inverter.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("energy_per_inverter.json"),
    )

    solarlog_connector = SolarLogConnector(
        "http://solarlog.com",
        True,
        "UTC",
        {0: True, 1: True, 2: False, 3: True},
    )

    data = await solarlog_connector.update_data()

    assert data == snapshot

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed


@pytest.mark.parametrize(
    ("status", "request_timeout", "error"),
    [
        (200, True, SolarLogConnectionError),
        (400, False, SolarLogUpdateError),
    ],
)
async def test_update_data_exceptions(
    responses: aioresponses,
    status: int,
    request_timeout: bool,
    error: SolarLogError,
) -> None:
    """Test update data with exceptions."""
    responses.post(
        "http://solarlog.com/getjp",
        timeout=request_timeout,
        status=status,
    )

    solarlog_connector = SolarLogConnector(
        "http://solarlog.com",
        True,
        "UTC",
        {0: True, 1: False, 2: False, 3: True},
    )
    await solarlog_connector.client.close()
    solarlog_connector.client.session = None  # type: ignore [assignment]

    with pytest.raises(error):  # type: ignore [call-overload]
        await solarlog_connector.update_data()

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed


async def test_update_data_with_data_exceptions(
    responses: aioresponses,
) -> None:
    """Test update data with exceptions due to data."""
    responses.post(
        "http://solarlog.com/getjp",
        body="",
    )
    responses.post(
        "http://solarlog.com/getjp",
        body='{"QUERY IMPOSSIBLE 000"}',
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("basic_data_during_update.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("basic_data_no_power.json"),
    )

    solarlog_connector = SolarLogConnector("http://solarlog.com")

    with pytest.raises(SolarLogUpdateError):
        await solarlog_connector.update_data()

    with pytest.raises(SolarLogUpdateError):
        await solarlog_connector.update_data()

    with pytest.raises(SolarLogUpdateError):
        await solarlog_connector.update_data()

    data = await solarlog_connector.update_data()

    assert data.usage == 0
    assert data.power_available == 0

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed


async def test_update_device_list(
    responses: aioresponses,
    snapshot: SnapshotAssertion
) -> None:
    """Test update device list."""

    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_list.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_data_1.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_data_2.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_data_3.json"),
    )
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_data_4.json"),
    )

    solarlog_connector = SolarLogConnector(
        "http://solarlog.com",
        True,
        "UTC",
        {0: True, 1: False, 2: False, 3: True},
    )

    await solarlog_connector.update_device_list()
    data = solarlog_connector.device_list

    assert data == snapshot

    assert solarlog_connector.device_name(0) == "Device 1"
    assert solarlog_connector.device_name(4) == ""

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed


async def test_enabled_devices(responses: aioresponses) -> None:
    """Test enabled devices."""
    responses.post(
        "http://solarlog.com/getjp",
        body=load_fixture("device_list.json"),
    )

    solarlog_connector = SolarLogConnector("http://solarlog.com",device_enabled={})
    assert solarlog_connector.device_enabled() == {}
    assert await solarlog_connector.update_device_list() == {}

    solarlog_connector.set_enabled_devices({0: True, 1: False, 2: False, 3: True})

    assert solarlog_connector.device_enabled(0)
    assert not solarlog_connector.device_enabled(1)
    assert solarlog_connector.device_enabled(3)

    await solarlog_connector.client.close()
    assert solarlog_connector.client.session.closed
