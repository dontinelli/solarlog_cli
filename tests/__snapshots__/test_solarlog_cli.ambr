# serializer version: 1
# name: test_update_data[asyncio]
  dict({
    'alternator_loss': 100,
    'capacity': 27.901119402985074,
    'consumption_ac': 3110,
    'consumption_day': 47819,
    'consumption_month': 697471,
    'consumption_total': 28385247,
    'consumption_year': 4218251,
    'consumption_yesterday': 94350,
    'efficiency': 96.65663657639585,
    'inverter_data': dict({
      0: dict({
        'consumption_year': 4227027.0,
        'current_power': 3170.0,
        'enabled': True,
        'name': '',
      }),
      1: dict({
        'consumption_year': 1920650.0,
        'current_power': None,
        'enabled': True,
        'name': '',
      }),
      2: dict({
        'consumption_year': None,
        'current_power': None,
        'enabled': False,
        'name': '',
      }),
      3: dict({
        'consumption_year': None,
        'current_power': 2816.0,
        'enabled': True,
        'name': '',
      }),
    }),
    'last_updated': datetime.datetime(2024, 8, 26, 14, 19, 45, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
    'power_ac': 2891,
    'power_available': -219,
    'power_dc': 2991,
    'production_year': 8852462,
    'self_consumption_year': 2071,
    'total_power': 10720,
    'usage': 107.5752334832238,
    'voltage_ac': 0,
    'voltage_dc': 492,
    'yield_day': 17490,
    'yield_month': 1179844,
    'yield_total': 55543544,
    'yield_year': 8843344,
    'yield_yesterday': 12955,
  })
# ---
# name: test_update_device_list[asyncio]
  dict({
    0: dict({
      'consumption_year': None,
      'current_power': None,
      'enabled': True,
      'name': 'Device 1',
    }),
    1: dict({
      'consumption_year': None,
      'current_power': None,
      'enabled': False,
      'name': 'Device 2',
    }),
    2: dict({
      'consumption_year': None,
      'current_power': None,
      'enabled': False,
      'name': 'Device 3',
    }),
    3: dict({
      'consumption_year': None,
      'current_power': None,
      'enabled': True,
      'name': 'Device 4',
    }),
  })
# ---
